File: tools/copyIfNewer/CopyIfNewer.cpp
#include "G3D/G3D.h"

using namespace G3D;

void printHelp();
bool isDirectory(const String& filename);
/** Adds a slash to a directory, if not already there. */
String maybeAddSlash(const String& in);
int main(int argc, char** argv);
bool excluded(bool exclusions, bool superExclusions, const String& filename);


void copyIfNewer
   (bool exclusions, 
    bool superExclusions, 
    String sourcespec,
    String destspec) {

	if (FileSystem::isDirectory(sourcespec)) {
		// Copy an entire directory.  Change the arguments so that
		// we copy the *contents* of the directory.

		sourcespec = maybeAddSlash(sourcespec);
		sourcespec = sourcespec + "*";
	}

	String path = filenamePath(sourcespec);

	Array<String> fileArray;
	Array<String> dirArray;

	FileSystem::getDirectories(sourcespec, dirArray);
	FileSystem::getFiles(sourcespec, fileArray);

	destspec = maybeAddSlash(destspec);

	if (FileSystem::exists(destspec, false) && ! FileSystem::isDirectory(destspec)) {
		printf("A file already exists named %s.  Target must be a directory.", 
			destspec.c_str());
		exit(-2);
	}
	FileSystem::createDirectory(destspec);

	for (int f = 0; f < fileArray.length(); ++f) {
		if (! excluded(exclusions, superExclusions, fileArray[f])) {
			String s = path + fileArray[f];
			String d = destspec + fileArray[f];
			if (FileSystem::isNewer(s, d)) {
				printf("copy %s %s\n", s.c_str(), d.c_str());
				FileSystem::copyFile(s, d);
			}
		}
	}

	// Directories just get copied; we don't check their dates.
	// Recurse into the directories
	for (int d = 0; d < dirArray.length(); ++d) {
		if (! excluded(exclusions, superExclusions, dirArray[d])) {
			copyIfNewer(exclusions, superExclusions, path + dirArray[d], destspec + dirArray[d]);
		}
	}
}



int main(int argc, char** argv) {

    initG3D();

    if (((argc == 2) && (String("--help") == argv[1])) || (argc < 3) || (argc > 4)) {
        printHelp();
        return -1;
    } else {
        bool exclusions = false;
        bool superExclusions = false;

        String s, d;
        if (String("--exclusions") == argv[1]) {
            exclusions = true;
            s = argv[2];
            d = argv[3];
        } else if (String("--super-exclusions") == argv[1]) {
            exclusions = true;
            superExclusions = true;
            s = argv[2];
            d = argv[3];
        } else {
            s = argv[1];
            d = argv[2];
        }

        copyIfNewer(exclusions, superExclusions, s, d);
    }
    
    return 0;
}


void printHelp() {
    printf("COPYIFNEWER\n\n");
    printf("SYNTAX:\n\n");
    printf(" copyifnewer [--help] [--exclusions | --super-exclusions] <source> <destdir>\n\n");
    printf("ARGUMENTS:\n\n");
    printf("  --exclusions  If specified, exclude CVS, svn, and ~ files. \n\n");
    printf("  --super-exclusions  If specified, exclude CVS, svn, ~, .ncb, .pyc, .sdf, .ncb, .suo Release, Debug, build, temp files. \n\n");
    printf("  source   Filename or directory name (trailing slash not required).\n");
    printf("           May include standard Win32 wild cards in the filename.\n");
    printf("  dest     Destination directory, no wildcards allowed.\n\n");
    printf("PURPOSE:\n\n");
    printf("Copies files matching the source specification to the dest if they\n");
    printf("do not exist in dest or are out of date (according to the file system).\n\n");
    printf("Compiled: " __TIME__ " " __DATE__ "\n"); 
}


String maybeAddSlash(const String& sourcespec) {
    if (sourcespec.length() > 0) {
        char last = sourcespec[sourcespec.length() - 1];
        if ((last != '/') && (last != ':') && (last != '\\')) {
            // Add a trailing slash
            return sourcespec + "/";
        }
    }
    return sourcespec;
}

bool excluded(bool exclusions, bool superExclusions, const String& filename) {
    
    if (exclusions) {
        if (filename[filename.length() - 1] == '~') {
            return true;
        } else if ((filename == "CVS") || (filename == "svn") || (filename == ".svn") || (filename == ".cvsignore")) {
            return true;
        }
    }

    if (superExclusions) {
        String f = toLower(filename);
        if ((f == "release") || 
            (f == "debug") ||
            (f == "build") ||
            (f == "graveyard") ||
            (f == "temp") ||
            endsWith(f, ".pyc") ||
            endsWith(f, ".sbr") ||
            endsWith(f, ".ncb") ||
            endsWith(f, ".opt") ||
            endsWith(f, ".bsc") ||
            endsWith(f, ".suo") ||
            endsWith(f, ".ncb") ||
            endsWith(f, ".sdf") ||
            endsWith(f, ".pch") ||
            endsWith(f, ".ilk") ||
            endsWith(f, ".pdb")) {

            return true;
        }
    }

    return false;
}

File: G3D10/bin/ice/copyifnewer.py
# copyifnewer.py
#
#
from __future__ import print_function

import re, string, glob
from .utils import *

_excludeDirPatterns = \
    ['^#',\
     '~$',\
     '^\.svn$',\
     '^\.git$',\
     '^CVS$', \
     '^Debug$', \
     '^Release$', \
     '^graveyard$', \
     '^tmp$', \
     '^temp$', \
     '^\.icompile-temp$', \
     '^\.ice-tmp$', \
     '^build$']
#'^\.',\

""" Regular expression patterns that will be excluded from copying by 
    copyIfNewer.
"""
_excludeFromCopyingPatterns =\
    ['\.ncb$', \
    '\.opt$', \
    '\.sdf$', \
    '\.ilk$', \
    '\.cvsignore$', \
    '^\.\#', \
    '\.pdb$', \
    '\.bsc$', \
    '^\.DS_store$', \
    '\.o$', \
    '\.pyc$', \
    '\.plg$', \
    '^#.*#$', \
    '^ice-stats\.csv$'\
    '~$', \
    '\.old$' \
    '^log.txt$', \
    '^stderr.txt$', \
    '^stdout.txt$', \
    '\.log$', \
    '\^.cvsignore$'] + _excludeDirPatterns

"""
  Regular expression patterns (i.e., directory and filename patterns) that are 
  excluded from the search for cpp files
"""
_cppExcludePatterns = ['^test$', '^tests$', '^#.*#$', '~$', '^old$'] + _excludeFromCopyingPatterns

"""
A regular expression matching files that should be excluded from copying.
"""
excludeFromCopying  = re.compile('|'.join(_excludeFromCopyingPatterns))


"""
Recursively copies all contents of source to dest 
(including source itself) that are out of date.  Does 
not copy files matching the excludeFromCopying patterns.

Returns a list of the files (if any were copied)

If actuallyCopy is false, doesn't actually copy the files, but still prints.

"""
def copyIfNewer(source, dest, echoCommands = True, echoFilenames = True, actuallyCopy = True):
    copiedFiles = []
    
    if source == dest:
        # Copying in place
        #print('copyIfNewer: Copying in place...nothing to do')
        1 # some statement

    elif ('*' in source) or ('?' in source):
        # expand wildcards
        for s in glob.glob(source):
            copiedFiles += copyIfNewer(s, dest, echoCommands, echoFilenames, actuallyCopy)

    else:
        if not os.path.exists(source):
            # Source does not exist
            print('copyIfNewer: Source (' + source + ') does not exist. Nothing to do')

        elif os.path.isdir(source):
            # Walk is a special iterator that visits all of the children recursively
            for dirpath, subdirs, filenames in os.walk(source):
                copiedFiles += _copyIfNewerVisit(len(source), dest, echoCommands, echoFilenames, actuallyCopy, dirpath, filenames, subdirs)

        else:
            # Source is a single file

            if (dest[-1] == '/') or (dest[-1] == '\\'):
                # Destination is a directory. Figure out the filename
                dest = pathConcat(dest, betterbasename(source))

            #dest = removeTrailingSlash(dest)
            
            if newer(source, dest):
                if echoCommands: 
                    colorPrint('cp ' + source + ' ' + dest, COMMAND_COLOR)
                elif echoFilenames:
                    print(source)
        
                if actuallyCopy:
                    shutil.copyfile(source, dest)
                    copiedFiles += [source]
            elif echoCommands:
                print(dest + ' is up to date with ' + source)

    return copiedFiles
    
#########################################################################
    
"""Helper for copyIfNewer."""
def _copyIfNewerVisit(prefixLen, rootDir, echoCommands, echoFilenames, actuallyCopy, sourceDirname, files, subdirs):

    if (excludeFromCopying.search(betterbasename(sourceDirname)) != None):
        # Don't recurse into subdirectories of excluded directories
        del subdirs[:]
        return []

    # Construct the destination directory name
    # by concatenating the root dir and source dir
    destDirname = pathConcat(rootDir, sourceDirname[prefixLen:])
    dirName     = betterbasename(destDirname)

    # Create the corresponding destination dir if necessary
    if actuallyCopy:
        mkdir(destDirname, echoCommands)

    copiedFiles = []
    # Iterate through the contents of this directory   
    for name in files:
        source = pathConcat(sourceDirname, name)

        if excludeFromCopying.search(name) == None:
            
            # Copy files if newer
            dest = pathConcat(destDirname, name)
            if (newer(source, dest)):
                if echoCommands:
                    colorPrint('cp ' + source + ' ' + dest, COMMAND_COLOR)
                elif echoFilenames: 
                    print(name)
                copiedFiles += [source]
                if actuallyCopy:
                    shutil.copy(source, dest)

    return copiedFiles
